### ISSUE TITLE
[UX] Add a help text indicator & tooltip for placing next to things that may be confusing

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#3707]
**Description of the need**
There are some interfaces in Backdrop that are daunting to new users, and confusing even to experienced users. On these kinds of interfaces, it can be hard to agree on any changes that would help clarify things without adding to the clutter or overwhelming nature of the page. (Examples: Layouts landing page, Views configuration page)

Our first instinct is often to add help text, but if the explanation is only for a specific part of the page, finding a place to add that text can be hard. And as usability studies have proven, the more text is on a page, the less likely people are to actually read it.

**Proposed solution**
I would like to propose a new user-interface pattern that we could introduce, and then use consistently throughout core, wherever we have a sufficiently complex or overwhelming interface that needs more explanation: 

An icon that's a question-mark in a circle, that provides a tooltip containing additional help text. When you click on (or hover over?) the `?` a modal could display showing more information about that part of the page -- in the case of this example, an explanation for the `Reorder` button and the `Menu settings` button.

<img width="2696" alt="Screen Shot 2019-04-23 at 4 39 02 PM" src="https://user-images.githubusercontent.com/397895/56622832-7cab5f00-65e6-11e9-8c75-13a7a6b74ef0.png">

(I think we should use an icon instead of the text `?` but in the interest of creating a mockup quickly I used text.)

**Alternatives that have been considered**

- renaming button labels
- addition of help text
- addition of more functionality

Is there a contributed module that accomplishes this? ...if so, then has this
been ported to Backdrop?: 

- yes, there is https://www.drupal.org/project/tooltip
- no, not ported to B

**Comment by olafgrabienski** [comment]
> Is there a contributed module that accomplishes this? ...if so, then has this
been ported to Backdrop?

Maybe of interest: https://backdropcms.org/project/beautytips

**Comment by klonos** [comment]
Very closely related: #1414

**Comment by klonos** [comment]
I like this idea 👍

**Comment by klonos** [comment]
...cross-posting this image from #1414:

![backdrop-help_text_below_vs_help_text_above_vs_help_icons](https://cloud.githubusercontent.com/assets/2423362/12004213/3e58c48a-ab9b-11e5-9bb2-00aeca712898.png)

**Comment by klonos** [comment]
...and also cross-linking #1403

**Comment by jenlampton** [comment]
IIRC, Beauty tips takes things like form element descriptions and puts them into tooltips. I generally find that a bit of a UX regression (though, it is pretty). 

We should look at how they handle the tooltip part though, as I still think there's likely things we could benefit from in that module :)

> ...cross-posting this image from

Yes! I want this :)

**Comment by Graham-72** [comment]
Yes, me too. It is interesting to read our discussions from three + years ago in those linked issues.

**Comment by klonos** [comment]
This was implemented in the Webform module (8.x - not sure about 7.x):

![Screen Shot 2019-09-21 at 6 45 35 pm](https://user-images.githubusercontent.com/2423362/65370804-1ebb9a00-dca0-11e9-8015-9d37b8c1d855.png)

**Comment by jenlampton** [comment]
See also https://github.com/backdrop/backdrop-issues/issues/1414

**Comment by stpaultim** [comment]
I just closed in issue #4680 in favor of this issue, because I think they are pretty much the same, except that one branded this feature as "Dragon Help," which I like. 

I considered closing #1414 in favor of this issue, but I think the suggestion is slightly different there. #1414 is specifically suggesting that we have module that replaces the "description" field with a "help" icon. I think that this issue MIGHT be suggesting something slightly different. This one might be suggesting that the "help" icon suppliment rather than replace the help icon.

Thoughts?

NOTE: This topic was discussed in the Design/UX meeting yesterday at the 25:00 mark: https://youtu.be/vVwy8UOZoD0

During discussion there was no clear decision. But, the overall theme seems to be that this might be best served by a contrib module at first. One of the fears is that we don't want to add a module that exposes "help" options, if there is not enough "help" text available and the feature does not get enough help text provided.

**Comment by jenlampton** [comment]
> I considered closing #1414 in favor of this issue,

We usually close newer issues in favor of older ones if they are the same thing (unless the older one had no activity) so my vote would be to close this one if we do decide to consolidate. (Even though they are slightly different, I could see merging the two)

**Comment by klonos** [comment]
Yeah, my understanding is that #1414 is about allowing people that feel that they don't need help in their face all the time, to basically have a way to "collapse" all existing help text on the site (**which has been added to form fields**) into a ? icon + a tooltip. This could be implemented as:
- a per-user setting (a checkbox in the user profile form)
- a site-wide setting that site owners could "force" for all users
- a site-wide setting that is enabled when forms are being viewed on narrower screens.

This issue here though is about introducing a way to allow help text to be added in the form of a ? icon + tooltip in various UI elements (i.e. not just form elements - things like submit and operations buttons etc.).

**Comment by bugfolder** [comment]
Totally agree with @jenlampton 's proposed solution. (Unaware of this issue, I suggested much the same thing in the Zulip chat.) The core idea of "keep information out of the user's way, but let them easily get to it." Also suggested including links to the documentation in the tooltip.

**Comment by bugfolder** [comment]
I looked at the two existing contrib modules mentioned above:

* [Tooltip](https://www.drupal.org/project/tooltip) is just an API function and relies on the user to sanitize. There would still need to be a layer to handle sanitization and consistency to provide tooltips broadly.
* [Beautytips](https://backdropcms.org/project/beautytips) is more functionally complete, but it only displays what's in the title attribute, so someone wanting to add this information to (say) a render array would need to do pretty low-level creation of the attribute. This seems particularly problematic if you want the tip to contain formatted information (e.g., lists, links to further docs).

So it seems like something more extensive might be needed. Ideally, the content of a tooltip should be able to contain all the complexity possible in formatted text.

What about the idea of adding another property to the form API? LIke, say, '#tooltip', along with enough support in the core form API to interpret it and generate the appropriate rendered HTML? That would be in addition to '#description', but with the distinction provided that '#description' is displayed by default, '#tooltip' is hidden by default. With both options available, developers could then have the option of making their descriptions briefer, and can be as wordy as they like in tooltips, and would allow for progressive adoption over time.

**Comment by klonos** [comment]
> The core idea of "keep information out of the user's way, but let them easily get to it."

@bugfolder you might also like this solution that (sorta) tries to mimic the WordPress expandable/collapsible big-chunk-of-help-text-yet-still-out-of-the-way solution, but tries to do it with what we currently have in core (fieldsets): #4868 ...I have a PR with a working sandbox that you can test and provide feedback re the UX 😉 

> Also suggested including links to the documentation in the tooltip.

Yup, my PR includes that too 😉 

> What about ... '#tooltip'... in addition to '#description', but with the distinction provided that '#description' is displayed by default, '#tooltip' is hidden by default.

I would be supportive of that 👍
